### DISCUSSION TITLE
valueAsNumber in Controllers

### DISCUSSION THREAD

**Discussion by AdiHefferLusha** [react-hook-form/react-hook-form#8068] [Q&A] [answered]
Hi all,
I'm using a material UI input, wrapped by a controller, and I would like to get its submitted data as a parsed number.
However, the valueAsNumber rule doesn't seem to work for controllers, as shown in this sandbox:
https://codesandbox.io/s/zealous-currying-90d5wf?file=/src/App.js
(type any number into both inputs and submit to see the controlled value is still a string)

I didn't open a bug, since I've read the documentation closely, and it doesn't explicitly say that the controller's rules include valueAsNumber.
My question is: Should valueAsNumber work on a controller? If not, is there another way to configure my controller to retrieve a number?

Thanks!

**Comment by Moshyfawn** [comment, marked as answer]
`Controller` doesn't pack the `valueAs` functionality by design.

In the `Controller` types _(omitted most of unrelated types for simplicity)_:
```ts
type UseControllerProps<...> = {
    ...
    rules?: Omit<RegisterOptions<TFieldValues, TName>, 'valueAsNumber' | 'valueAsDate' | 'setValueAs' | 'disabled'>;
    ....
};
```

You can simply parse your input's value before passing it to the `Controller`'s `onChange` handler
```diff
<Controller
-    rules={{ valueAsNumber: true }}
      render={({ field }) => (
          <input
-             onChange={field.onChange}
+             onChange={(event) => field.onChange(+event.target.value)}
              type="number"
          />
  )}
/>
```

**Reply by AdiHefferLusha** [reply]
Thanks for the quick reply!
I'll try it.
Is there a reason the valueAs functions are not included, or is this just a matter of feature request?

**Reply by Moshyfawn** [reply]
`Controller` API is designed for controlled inputs, where you fully control the value and all the value handling yourself. So, you're already in control of parsing and formatting your value and how it gets into the RHF state. There's no need for RHF to additionally do anything with the value.

P.S. A bunch of instances of **control** word in my sentences, but that's the main gist of it :P

**Reply by eusahn** [reply]
Doesn't work for negative numbers.

**Reply by Moshyfawn** [reply]
> Doesn't work for negative numbers.

Care to elaborate? [Works fine for me](https://codesandbox.io/s/green-silence-7vxexr?file=%2Fsrc%2FApp.tsx)

**Reply by eusahn** [reply]
Try starting a number off with a dash. Also in the sandbox, you can't remove that 0. The problem is onChange is ran.. on change so trying to convert the dash itself will give you a NaN. You'll need to regex check the string for a number, then return the the actual converted value.

**Reply by Moshyfawn** [reply]
Valid point 👍 I guess, for controlled inputs to support the negative values you'd have to not parse the string value _(so, store it as `string` in RHF form data)_ and provide additional `validate` callback / validation schema logic to avoid "bad" values and keep the user input as is

**Reply by bluebill1049** [reply]
<img width="917" alt="Screen Shot 2022-07-06 at 8 56 20 am" src="https://user-images.githubusercontent.com/10513364/177430383-37f6d5fb-8de9-435d-aa11-da9c2a4a0f43.png">

take a look at this section above.

https://react-hook-form.com/api/usecontroller

there is a codesandbox as well: https://codesandbox.io/s/usecontroller-own-state-wncet2?file=/src/App.tsx

**Reply by mkosir** [reply]
@bluebill1049 Thanks for provided example 👍 .
I agree that responsibility should be on the UI component side, but I think the majority of confusion for devs comes as `valueAsNumber` parameter is available in `register` function (why does RHF handles it that case).

**Reply by bluebill1049** [reply]
Because `register` is applied to the uncontrolled components, if there is a massive demand for such a feature, we will reconsider it next year.

**Reply by bluebill1049** [reply]
here is related feature request: https://github.com/react-hook-form/react-hook-form/projects/1

**Reply by AmitNeuhaus** [reply]
@bluebill1049 
Can you explain how will your sandbox work when we want to reset the value of the controlled input?
We will have to reset both the UI state value (which is seperated from the field value), but how can we listen that  the field.value reseted and call setUiValue('')?

**Reply by takanuva15** [reply]
I just found this thread while trying to look up the same functionality myself lol. Incidentally, I was also confused why `valueAsNumber` is available for `register` but not for `Controller`. 

I think adding the `valueAs` props for Controllers would be great, and we can just put a disclaimer in the docs that the relevant transformation is applied after the invocation of the onChange method (in case someone tries to put their own custom event function like in the example).

**Reply by gabriellemos** [reply]
Had some issues with typescript because RHF has a different definition for `onChange`, to solve it I used the following instead of a plain old html input. 

```tsx
import type { ComponentPropsWithoutRef } from 'react'
import React from 'react'
import type { ControllerRenderProps } from 'react-hook-form'

type NumberInputProps = Pick<ControllerRenderProps, 'onChange'> &
  Omit<ComponentPropsWithoutRef<'input'>, 'onChange'>

const NumberInput = ({ onChange, ...props }: NumberInputProps) => {
  return (
    <input
      {...props}
      onChange={(event) => onChange?.(parseInt(event.target.value, 10))}
      type="number"
    />
  )
}

export default NumberInput

```

**Reply by sacru2red** [reply]
how to vote https://github.com/react-hook-form/react-hook-form/issues/6841

**Comment by clemente-xyz** [comment]
I think is not good designed and is a not scalable workaround. numerical inputs are numerical, they should be returned as so by react-hook-form. I think is a bad implementation what they did there. Hope they fix it on a future :slightly_smiling_face:

**Reply by bluebill1049** [reply]
Not here to argue right or wrong, just to provide information so developers don't get confused about such.

> numerical inputs are numerical

That's incorrect.

Please refer to the HTML spec: https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement

![Screenshot 2022-12-29 at 9 36 03 am](https://user-images.githubusercontent.com/10513364/209881307-afc4afca-1e8d-48f6-8e0b-fcf53fc1a1ff.png)

**Reply by cfaester** [reply]
They are strings and should not be treated as numerical. If you were to use `valueAsNumber` you'd find that a fractional number, would falsely trigger `onChange` with incorrect values. That is to say the `valueAsNumber` of "2." (notice the dot), is just "2". If you in turn use this as your value in your controlled component, the user would never be allowed to finish typing their number.

**Comment by milon27** [comment]
please add this feature for date and number it will be very helpfull

**Comment by msageryd** [comment]
I'll just chime in and add my 2 cents.

Way back I had a discussion with @bluebill1049 about when to cast form values to actual values. I got the advice to keep all form input as strings and handle the conversion outside of the form, i.e. when setting defaultValues and when submitting.

I did as I was adviced, and I'm very happy with this. My inputs behave much better this way than if I had tried to enforce some realtime casting.

This is what I do:
- my initial formData has correct datatypes. This is what is stored in my database.
- just before sending my formData to the form via defaultValues I convert everything to what I call "componentData". This is mostly strings, but checkboxes are booleans. This also implies that all empty values are converted to empty strings, which solves the problem with resetting the form.
- my inputFields has onChange handlers which ensures only valid format is inserted, i.e. only numeric, etc
- I actually allow both comma and dot as decimal separator. This helps with devices set to different input languages, which might give you the wrong decimal character on the keyboard.
- onSubmit converts everything back to correct datatypes (typedData), including replacing any commas with dots in numerical fields

Here is how I convert from "componentData", i.e. strings to correct datatypes.
<img width="473" alt="image" src="https://github.com/react-hook-form/react-hook-form/assets/4483085/af516cc4-7299-43a8-bd92-c2e4574f01fc">

Conversion from "typedData" to "componentData":
<img width="552" alt="image" src="https://github.com/react-hook-form/react-hook-form/assets/4483085/6bd4fea9-c738-4659-927a-197ce0faa92e">


`inputPattern` is checked in onChange. If something not matching inputPattern is entered, the `value` is reverted to `lastValue`. This way it's easy to use strings as input, and also be sure that no invalid data is entered.

I found it convenient to separate `componentTypes` from `dataTypes`.

I'm sorry if my post does not answer the question. I'm just so happy with my solution, so I thought I'd share it.

**Reply by dlsso** [reply]
I think this makes the most sense. To clarify the implementation, if I am currently trying to use numbers and have something like this:

```typescript
export type FormInputs = {
  title: string;
  numberField: number;
};

useForm<FormInputs>({
    values: {
      title: data.title || '',
      numberField: data.numberField || 0,
    },
})
    
// Data received from the submitted form is typed like this
const handleSubmit = (data: FormInputs) => {...}
```

You would change the type in both places to reflect the actual types received, and a new type (not pictured) that still has the number type would be used for the incoming data before it is converted.

```typescript
export type componentTypes = {
  title: string;
  numberField: string; // Main change here
};

useForm<componentTypes>({
    values: {
      title: data.title || '',
      numberField: data.numberField || '0', // And small change here
    },
})
    
// Data received from the submitted form is typed like this
const handleSubmit = (data: componentTypes) => {...}
```

Am I understanding correctly?

**Comment by OlegTar** [comment]
I have object 
```
{
     date: Date;
}
```

I use 
onChange(convertToDate(v))

but in handleSubmit type of date is always string;
